lib/racecar/liveness_probe.rb

```diff
@@ -0,0 +1,73 @@
+require "fileutils"
+
+module Racecar
+  class LivenessProbe
+    def initialize(message_bus, file_path, max_interval)
+      @message_bus = message_bus
+      @file_path = file_path
+      @max_interval = max_interval
+      @subscribers = []
+    end
+
+    attr_reader :message_bus, :file_path, :max_interval, :subscribers
+    private     :message_bus, :file_path, :max_interval, :subscribers
+
+    def check_liveness_within_interval!
+      liveness_event_within_interval? or
+        raise "Racecar healthcheck failed: No liveness within interval #{max_interval}s. Last liveness at #{last_liveness_event_at}, #{elapsed_since_liveness_event} seconds ago."
+    end
+
+    def liveness_event_within_interval?
+      elapsed_since_liveness_event < max_interval
+    end
+
+    def install
+      unless file_path && file_writeable?
+        raise(
+          "Liveness probe configuration error: `liveness_probe_file_path` must be set to a writable file path.\n" \
+            "  Set `RACECAR_LIVENESS_PROBE_FILE_PATH` and `RACECAR_LIVENESS_MAX_INTERVAL` environment variables."
+        )
+      end
+
+      subscribers << message_bus.subscribe("start_main_loop.racecar") do
+        touch_liveness_file
+      end
+
+      subscribers = message_bus.subscribe("shut_down.racecar") do
+        delete_liveness_file
+      end
+
+      nil
+    end
+
+    def uninstall
+      subscribers.each { |s| message_bus.unsubscribe(s) }
+    end
+
+    private
+
+    def elapsed_since_liveness_event
+      Time.now - last_liveness_event_at
```

Comment:
There _is_ an error scenario here where the local clock jumps. Should we consider using `Process.clock_gettime(Process::CLOCK_MONOTONIC)` and storing the values in the files instead of relying on `mtime` and `Time.now`? Considering that a clock jump might actually cause the kpods to be terminated.

Reply:
I always have the same thought when I see `Time.now`!

In the case of forward clock jump a single check could fail, but the next loop will will correct the situation.

I don't think killing a process on one liveness check failure is normal so I didn't add the complexity. If we want to support that use-case then np.

Reply:
You're right, it's probably not something that could trigger a Pod termination. We can always add it in later on :+1: